File: lib/models/billing.dart
import 'package:flutter/material.dart';

enum BillingStatus {
  pending,
  paid,
  overdue,
  cancelled,
  refunded,
}

enum PaymentMethod {
  creditCard,
  bankTransfer,
  cash,
  other,
}

class Billing {
  final String id;
  final String businessId;
  final String subscriptionId;
  final DateTime billingDate;
  final DateTime dueDate;
  final DateTime? paidDate;
  final BillingStatus status;
  final double amount;
  final double taxAmount;
  final double totalAmount;
  final PaymentMethod? paymentMethod;
  final String? paymentReference;
  final int staffCount;
  final String planName;
  final double pricePerStaff;
  final String? notes;
  final Map<String, dynamic>? metadata;

  const Billing({
    required this.id,
    required this.businessId,
    required this.subscriptionId,
    required this.billingDate,
    required this.dueDate,
    this.paidDate,
    required this.status,
    required this.amount,
    required this.taxAmount,
    required this.totalAmount,
    this.paymentMethod,
    this.paymentReference,
    required this.staffCount,
    required this.planName,
    required this.pricePerStaff,
    this.notes,
    this.metadata,
  });

  factory Billing.fromJson(Map<String, dynamic> json) {
    return Billing(
      id: json['id'],
      businessId: json['businessId'],
      subscriptionId: json['subscriptionId'],
      billingDate: DateTime.parse(json['billingDate']),
      dueDate: DateTime.parse(json['dueDate']),
      paidDate: json['paidDate'] != null ? DateTime.parse(json['paidDate']) : null,
      status: BillingStatus.values.firstWhere(
        (e) => e.name == json['status'],
        orElse: () => BillingStatus.pending,
      ),
      amount: json['amount'].toDouble(),
      taxAmount: json['taxAmount'].toDouble(),
      totalAmount: json['totalAmount'].toDouble(),
      paymentMethod: json['paymentMethod'] != null
          ? PaymentMethod.values.firstWhere(
              (e) => e.name == json['paymentMethod'],
              orElse: () => PaymentMethod.other,
            )
          : null,
      paymentReference: json['paymentReference'],
      staffCount: json['staffCount'],
      planName: json['planName'],
      pricePerStaff: json['pricePerStaff'].toDouble(),
      notes: json['notes'],
      metadata: json['metadata'],
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'businessId': businessId,
      'subscriptionId': subscriptionId,
      'billingDate': billingDate.toIso8601String(),
      'dueDate': dueDate.toIso8601String(),
      'paidDate': paidDate?.toIso8601String(),
      'status': status.name,
      'amount': amount,
      'taxAmount': taxAmount,
      'totalAmount': totalAmount,
      'paymentMethod': paymentMethod?.name,
      'paymentReference': paymentReference,
      'staffCount': staffCount,
      'planName': planName,
      'pricePerStaff': pricePerStaff,
      'notes': notes,
      'metadata': metadata,
    };
  }

  String get statusDisplayName {
    switch (status) {
      case BillingStatus.pending:
        return '待付款';
      case BillingStatus.paid:
        return '已付款';
      case BillingStatus.overdue:
        return '逾期未付';
      case BillingStatus.cancelled:
        return '已取消';
      case BillingStatus.refunded:
        return '已退款';
    }
  }

  Color get statusColor {
    switch (status) {
      case BillingStatus.pending:
        return const Color(0xFFFF9800);
      case BillingStatus.paid:
        return const Color(0xFF4CAF50);
      case BillingStatus.overdue:
        return const Color(0xFFF44336);
      case BillingStatus.cancelled:
        return const Color(0xFF9E9E9E);
      case BillingStatus.refunded:
        return const Color(0xFF2196F3);
    }
  }

  String get paymentMethodDisplayName {
    if (paymentMethod == null) return '未設定';
    switch (paymentMethod!) {
      case PaymentMethod.creditCard:
        return '信用卡';
      case PaymentMethod.bankTransfer:
        return '銀行轉帳';
      case PaymentMethod.cash:
        return '現金';
      case PaymentMethod.other:
        return '其他';
    }
  }

  bool get isPaid => status == BillingStatus.paid;
  bool get isPending => status == BillingStatus.pending;
  bool get isOverdue => status == BillingStatus.overdue;

  int get daysUntilDue {
    final now = DateTime.now();
    if (dueDate.isBefore(now)) return 0;
    return dueDate.difference(now).inDays;
  }

  int get daysOverdue {
    final now = DateTime.now();
    if (dueDate.isAfter(now)) return 0;
    return now.difference(dueDate).inDays;
  }

  String get formattedBillingDate => '${billingDate.year}/${billingDate.month.toString().padLeft(2, '0')}/${billingDate.day.toString().padLeft(2, '0')}';
  String get formattedDueDate => '${dueDate.year}/${dueDate.month.toString().padLeft(2, '0')}/${dueDate.day.toString().padLeft(2, '0')}';
  String get formattedPaidDate => paidDate != null ? '${paidDate!.year}/${paidDate!.month.toString().padLeft(2, '0')}/${paidDate!.day.toString().padLeft(2, '0')}' : '';

  String get formattedAmount => 'NT\$ ${amount.toInt()}';
  String get formattedTaxAmount => 'NT\$ ${taxAmount.toInt()}';
  String get formattedTotalAmount => 'NT\$ ${totalAmount.toInt()}';

  Billing copyWith({
    String? id,
    String? businessId,
    String? subscriptionId,
    DateTime? billingDate,
    DateTime? dueDate,
    DateTime? paidDate,
    BillingStatus? status,
    double? amount,
    double? taxAmount,
    double? totalAmount,
    PaymentMethod? paymentMethod,
    String? paymentReference,
    int? staffCount,
    String? planName,
    double? pricePerStaff,
    String? notes,
    Map<String, dynamic>? metadata,
  }) {
    return Billing(
      id: id ?? this.id,
      businessId: businessId ?? this.businessId,
      subscriptionId: subscriptionId ?? this.subscriptionId,
      billingDate: billingDate ?? this.billingDate,
      dueDate: dueDate ?? this.dueDate,
      paidDate: paidDate ?? this.paidDate,
      status: status ?? this.status,
      amount: amount ?? this.amount,
      taxAmount: taxAmount ?? this.taxAmount,
      totalAmount: totalAmount ?? this.totalAmount,
      paymentMethod: paymentMethod ?? this.paymentMethod,
      paymentReference: paymentReference ?? this.paymentReference,
      staffCount: staffCount ?? this.staffCount,
      planName: planName ?? this.planName,
      pricePerStaff: pricePerStaff ?? this.pricePerStaff,
      notes: notes ?? this.notes,
      metadata: metadata ?? this.metadata,
    );
  }
} 

File: backend/internal/models/billing.go
package models

import (
	"encoding/json"
	"time"

	"gorm.io/gorm"
)

// BillingStatus constants
const (
	BillingStatusPending   = "pending"
	BillingStatusPaid      = "paid"
	BillingStatusOverdue   = "overdue"
	BillingStatusCancelled = "cancelled"
	BillingStatusRefunded  = "refunded"
)

// PaymentMethod constants
const (
	PaymentMethodCreditCard   = "creditCard"
	PaymentMethodBankTransfer = "bankTransfer"
	PaymentMethodCash         = "cash"
	PaymentMethodOther        = "other"
)

// Billing represents the billing model
type Billing struct {
	ID               string         `gorm:"primaryKey;type:varchar(36)" json:"id"`
	BusinessID       string         `gorm:"type:varchar(36);not null;index" json:"businessId"`
	SubscriptionID   string         `gorm:"type:varchar(36);not null;index" json:"subscriptionId"`
	BillingDate      time.Time      `gorm:"not null" json:"billingDate"`
	DueDate          time.Time      `gorm:"not null" json:"dueDate"`
	PaidDate         *time.Time     `json:"paidDate,omitempty"`
	Status           string         `gorm:"type:varchar(20);default:'pending'" json:"status"`
	Amount           float64        `gorm:"type:decimal(10,2);not null" json:"amount"`
	TaxAmount        float64        `gorm:"type:decimal(10,2);default:0" json:"taxAmount"`
	TotalAmount      float64        `gorm:"type:decimal(10,2);not null" json:"totalAmount"`
	PaymentMethod    *string        `gorm:"type:varchar(20)" json:"paymentMethod,omitempty"`
	PaymentReference *string        `gorm:"type:varchar(100)" json:"paymentReference,omitempty"`
	StaffCount       int            `gorm:"not null" json:"staffCount"`
	PlanName         string         `gorm:"type:varchar(100);not null" json:"planName"`
	PricePerStaff    float64        `gorm:"type:decimal(10,2);not null" json:"pricePerStaff"`
	Notes            *string        `gorm:"type:text" json:"notes,omitempty"`
	Metadata         *string        `gorm:"type:json" json:"metadata,omitempty"` // JSON string
	CreatedAt        time.Time      `json:"created_at"`
	UpdatedAt        time.Time      `json:"updated_at"`
	DeletedAt        gorm.DeletedAt `gorm:"index" json:"-"`

	// Relationships
	Business     *Business     `gorm:"foreignKey:BusinessID" json:"business,omitempty"`
	Subscription *Subscription `gorm:"foreignKey:SubscriptionID" json:"subscription,omitempty"`
}

// TableName specifies the table name for Billing model
func (Billing) TableName() string {
	return "billings"
}

// BeforeCreate hook to generate UUID for billing ID
func (b *Billing) BeforeCreate(tx *gorm.DB) error {
	if b.ID == "" {
		b.ID = generateUUID()
	}
	return nil
}

// CreateBillingRequest represents the create billing request payload
type CreateBillingRequest struct {
	BusinessID     string                 `json:"businessId" binding:"required"`
	SubscriptionID string                 `json:"subscriptionId" binding:"required"`
	BillingDate    string                 `json:"billingDate" binding:"required"` // YYYY-MM-DD format
	DueDate        string                 `json:"dueDate" binding:"required"`     // YYYY-MM-DD format
	Amount         float64                `json:"amount" binding:"required,min=0"`
	TaxAmount      float64                `json:"taxAmount" binding:"omitempty,min=0"`
	TotalAmount    float64                `json:"totalAmount" binding:"required,min=0"`
	StaffCount     int                    `json:"staffCount" binding:"required,min=1"`
	PlanName       string                 `json:"planName" binding:"required"`
	PricePerStaff  float64                `json:"pricePerStaff" binding:"required,min=0"`
	Notes          string                 `json:"notes,omitempty"`
	Metadata       map[string]interface{} `json:"metadata,omitempty"`
}

// UpdateBillingRequest represents the update billing request payload
type UpdateBillingRequest struct {
	Status           string                 `json:"status,omitempty"`
	PaidDate         string                 `json:"paidDate,omitempty"` // YYYY-MM-DD format
	PaymentMethod    string                 `json:"paymentMethod,omitempty"`
	PaymentReference string                 `json:"paymentReference,omitempty"`
	Notes            string                 `json:"notes,omitempty"`
	Metadata         map[string]interface{} `json:"metadata,omitempty"`
}

// ToModel converts CreateBillingRequest to Billing model
func (r *CreateBillingRequest) ToModel() (*Billing, error) {
	// Parse billing date
	billingDate, err := time.Parse("2006-01-02", r.BillingDate)
	if err != nil {
		return nil, err
	}

	// Parse due date
	dueDate, err := time.Parse("2006-01-02", r.DueDate)
	if err != nil {
		return nil, err
	}

	billing := &Billing{
		BusinessID:     r.BusinessID,
		SubscriptionID: r.SubscriptionID,
		BillingDate:    billingDate,
		DueDate:        dueDate,
		Status:         BillingStatusPending,
		Amount:         r.Amount,
		TaxAmount:      r.TaxAmount,
		TotalAmount:    r.TotalAmount,
		StaffCount:     r.StaffCount,
		PlanName:       r.PlanName,
		PricePerStaff:  r.PricePerStaff,
	}

	if r.Notes != "" {
		billing.Notes = &r.Notes
	}

	// Convert metadata to JSON string if provided
	if r.Metadata != nil {
		metadataJSON, err := json.Marshal(r.Metadata)
		if err != nil {
			return nil, err
		}
		metadataStr := string(metadataJSON)
		billing.Metadata = &metadataStr
	}

	return billing, nil
}

// IsPaid returns true if the billing is paid
func (b *Billing) IsPaid() bool {
	return b.Status == BillingStatusPaid
}

// IsPending returns true if the billing is pending
func (b *Billing) IsPending() bool {
	return b.Status == BillingStatusPending
}

// IsOverdue returns true if the billing is overdue
func (b *Billing) IsOverdue() bool {
	return b.Status == BillingStatusOverdue
}

// DaysUntilDue returns the number of days until the billing is due
func (b *Billing) DaysUntilDue() int {
	now := time.Now()
	if b.DueDate.Before(now) {
		return 0
	}
	return int(b.DueDate.Sub(now).Hours() / 24)
}

// DaysOverdue returns the number of days the billing is overdue
func (b *Billing) DaysOverdue() int {
	now := time.Now()
	if b.DueDate.After(now) {
		return 0
	}
	return int(now.Sub(b.DueDate).Hours() / 24)
}
